### ISSUE TITLE
Can't reset password

### ISSUE THREAD

**Issue by edneymatias** [zammad/zammad#1305]
### Infos:

* Used Zammad version: 1.5.0
* Used Zammad installation source: package
* Operating system: Debian
* Browser + version: Chrome 59.0.3071.115

### Expected behavior:

* Tha password was reset successfully

### Actual behavior:

* {"error":"Argon2::ArgonHashFail"}

### Steps to reproduce the behavior:

* I used the login form to get a password reset email but when I try to do so, Zammad shows me the screen below. The informed password was "Alfabeta1" without the quotation marks. By the way, I'm pretty sure I know my actual password since I always use the same in my tests. 

![image](https://user-images.githubusercontent.com/22533602/28674931-ba12adc2-72bc-11e7-89d2-a8deb233dfef.png)

In production.log I see:

`[2017-07-27T14:25:38.476130 #10961] ERROR -- : Argon2::ArgonHashFail (Argon2::ArgonHashFail)
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/argon2-1.1.1/lib/argon2/ffi_engine.rb:59:in `block in hash_argon2i_encode'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/argon2-1.1.1/lib/argon2/ffi_engine.rb:55:in `initialize'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/argon2-1.1.1/lib/argon2/ffi_engine.rb:55:in `new'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/argon2-1.1.1/lib/argon2/ffi_engine.rb:55:in `hash_argon2i_encode'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/argon2-1.1.1/lib/argon2.rb:21:in `create'
/opt/zammad/lib/password_hash.rb:10:in `crypt'
/opt/zammad/app/models/user.rb:984:in `ensure_password'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:432:in `block in make_lambda'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:164:in `block in halting'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:504:in `block in call'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:504:in `each'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:504:in `call'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:92:in `__run_callbacks__'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.7.1/lib/active_support/callbacks.rb:778:in `_run_validation_callbacks'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activemodel-4.2.7.1/lib/active_model/validations/callbacks.rb:113:in `run_validations!'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activemodel-4.2.7.1/lib/active_model/validations.rb:338:in `valid?'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/validations.rb:58:in `valid?'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/validations.rb:83:in `perform_validations'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/validations.rb:37:in `save'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/attribute_methods/dirty.rb:21:in `save'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:286:in `block (2 levels) in save'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:351:in `block in with_transaction_returning_status'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract/database_statements.rb:211:in `transaction'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:220:in `transaction'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:348:in `with_transaction_returning_status'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:286:in `block in save'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:301:in `rollback_active_record_state!'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:285:in `save'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/persistence.rb:252:in `block in update'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:351:in `block in with_transaction_returning_status'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `block in transaction'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract/transaction.rb:184:in `within_new_transaction'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `transaction'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:220:in `transaction'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/transactions.rb:348:in `with_transaction_returning_status'
/opt/zammad/vendor/bundle/ruby/2.3.0/gems/activerecord-4.2.7.1/lib/active_record/persistence.rb:250:in `update'
/opt/zammad/app/models/user.rb:557:in `password_reset_via_token'`

Thanks in advance.

**Comment by martini** [comment]
Hi @edneymatias 

do you have installed "argon2" package on Debian? 

Feedback would be nice!

 -Martin

**Comment by edneymatias** [comment]
Hi Martin!

`root@helpdesk:/opt/zammad/log# dpkg -l | grep argon`
`root@helpdesk:/opt/zammad/log#`

Nop!

Thanks

**Comment by edneymatias** [comment]
Martin,

I deleted my previous feedback because I thought maybe you were talking about some rails package but I checked that too. Now by your answer, I'm sure you were really talking about Debian package. But there's no available package named argon2 on my distribution.

```
root@helpdesk:/opt/zammad/bin# apt-get install argon2
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package argon2
root@helpdesk:/opt/zammad/bin#
```

It seems I have to get it from sources on Debian 8, I can't do it right now, but will try later.

Thanks.

**Comment by martini** [comment]
> It seems I have to get it from sources on Debian 8, I can't do it right now, but will try later.

No no. Zammad runs on default Debian 8.

**Comment by edneymatias** [comment]
But it seems that an argon2 Debian package is available only from Stretch onwards.

**Comment by martini** [comment]
I'm not familiar with debian, but https://packages.debian.org/search?keywords=argon2 -> it should be available, right?

**Comment by edneymatias** [comment]
> I'm not familiar with debian, but https://packages.debian.org/search?keywords=argon2 -> it should be available, right?

That's because you have to limit the suite that you want to seein the package availability. At the top of the page, you can adjust that. Try jessie for example.

By the way, when I try apt-get update I get:

```
Get:20 https://deb.packager.io jessie/stable Translation-en                                                                                                                                                                                          
Ign https://deb.packager.io jessie/stable Translation-en                                                                                                                                                                                             
Fetched 148 kB in 8s (18.2 kB/s)                                                                                                                                                                                                                     
W: Failed to fetch https://deb.packager.io/gh/zammad/zammad/dists/jessie/stable/binary-amd64/Packages  Hash Sum mismatch
E: Some index files failed to download. They have been ignored, or old ones used instead.
```

Is that supposed to happen?

Thanks.

**Comment by martini** [comment]
> Is that supposed to happen?

No. The repo file format has changed, see #1294 (you need to update /etc/apt/sources.list.d/zammad.list)

Then it's working well again.

**Comment by martini** [comment]
JFI: I'll try it by my self on debian8 later. Normall it works just fine!

**Comment by thorsteneckel** [comment]
Where are you hosting? There are some issues with e.g. DigitalOcean. See #966. I wasn't able to reproduce it (on DigitalOcean) but would be happy to take a look.

**Comment by edneymatias** [comment]
> No. The repo file format has changed, see #1294 (you need to update /etc/apt/sources.list.d/zammad.list)

Great, done!

> Where are you hosting? There are some issues with e.g. DigitalOcean. See #966. I wasn't able to reproduce it (on DigitalOcean) but would be happy to take a look.

I'm running on Google Cloud which reminds me that I had an instance running a PHP application that used to get killed mysteriously but after a reboot, it used to work. So after a reboot of the Zammad instance, guess what? It worked!

I found that in the PHP case it was a low memory problem that went away when I raised the memory availability.

I'm running on an instance way below the hardware requirements and I think that the fresh start let enough memory for the hash computation. I'm just doing a broad test for now.

I think the case is closed.

Thank you all! You are a great team!
